src/stepFunctions/commands/visualizeStateMachine.ts

```diff
@@ -29,257 +29,341 @@ export interface messageObject {
-                'There was an error rendering State Machine Graph, check logs for details.'
+            return this.createNewVisualization(textDocument)
+        } catch (err) {
+            vscode.window.showInformationMessage(
+                localize(
+                    'AWS.stepfunctions.visualisation.errors.rendering',
+                    'There was an error rendering State Machine Graph, check logs for details.'
+                )
             )
-        )
 
-        logger.debug('Unable to setup webview panel.')
-        logger.error(err as Error)
+            logger.debug('Unable to setup webview panel.')
+            logger.error(err as Error)
+        }
+
+        return
     }
 
-    return
-}
+    private createNewVisualization(textDocument: vscode.TextDocument): vscode.WebviewPanel | void {
+        const newVisualization = new AslVisualization(textDocument)
+        this.managedVisualizations.add(newVisualization)
 
-async function setupWebviewPanel(
-    documentUri: vscode.Uri,
-    textDocument: vscode.TextDocument
-): Promise<vscode.WebviewPanel> {
-    const logger: Logger = getLogger()
-    let isPanelDisposed = false
-
-    // Create and show panel
-    const panel = vscode.window.createWebviewPanel(
-        'stateMachineVisualization',
-        makeWebviewTitle(documentUri),
-        {
-            preserveFocus: true,
-            viewColumn: vscode.ViewColumn.Beside,
-        },
-        {
-            enableScripts: true,
-            localResourceRoots: [
-                ext.visualizationResourcePaths.localWebviewScriptsPath,
-                ext.visualizationResourcePaths.visualizationLibraryCachePath,
-                ext.visualizationResourcePaths.stateMachineCustomThemePath,
-            ],
-            retainContextWhenHidden: true,
-        }
-    )
+        newVisualization.onVisualizationDispose(() => {
+            this.managedVisualizations.delete(newVisualization)
+        })
 
-    async function sendUpdateMessage(updatedTextDocument: vscode.TextDocument) {
-        const isValid = await isDocumentValid(updatedTextDocument)
+        return newVisualization.getPanel()
+    }
 
-        if (isPanelDisposed || !panel.webview) {
-            return
+    private getExistingVisualization(uriToFind: vscode.Uri): AslVisualization | void {
+        for (const vis of this.managedVisualizations) {
+            if (vis.documentURI.path === uriToFind.path) {
+                return vis
+            }
         }
 
-        logger.debug('Sending update message to webview.')
+        return
+    }
+}
 
-        panel.webview.postMessage({
-            command: 'update',
-            stateMachineData: updatedTextDocument.getText(),
-            isValid,
-        })
+class AslVisualization {
+    public readonly documentURI: vscode.Uri
+    public readonly webviewPanel: vscode.WebviewPanel
+    public onVisualizationDisposeEmitter = new vscode.EventEmitter<void>()
```

Comment:
onVisualizationDisposeEmitter should be private, and can be readonly. You wouldn't want code outside this class triggering (emitting) the event